File: app/src/main/java/com/example/personaltaskmanager/features/task_manager/viewmodel/TaskViewModel.java
package com.example.personaltaskmanager.features.task_manager.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.example.personaltaskmanager.features.task_manager.data.model.Task;
import com.example.personaltaskmanager.features.task_manager.data.repository.TaskRepository;
import com.example.personaltaskmanager.features.task_manager.domain.usecase.AddTaskUseCase;
import com.example.personaltaskmanager.features.task_manager.domain.usecase.DeleteTaskUseCase;
import com.example.personaltaskmanager.features.task_manager.domain.usecase.GetTasksUseCase;
import com.example.personaltaskmanager.features.task_manager.domain.usecase.GetTasksByDateUseCase;

import java.util.List;

/**
 * TaskViewModel quản lý dữ liệu Task theo mô hình MVVM.
 * Giữ nguyên code cũ, chỉ bổ sung logic deadline + getTasksByDate().
 */
public class TaskViewModel extends AndroidViewModel {

    private final TaskRepository repository;
    private final GetTasksUseCase getTasksUseCase;
    private final AddTaskUseCase addTaskUseCase;
    private final DeleteTaskUseCase deleteTaskUseCase;

    // NEW — use case lấy task theo ngày
    private final GetTasksByDateUseCase getTasksByDateUseCase;

    private final LiveData<List<Task>> allTasksLiveData;

    public TaskViewModel(@NonNull Application application) {
        super(application);

        repository = new TaskRepository(application);
        getTasksUseCase = new GetTasksUseCase(repository);
        addTaskUseCase = new AddTaskUseCase(repository);
        deleteTaskUseCase = new DeleteTaskUseCase(repository);

        getTasksByDateUseCase = new GetTasksByDateUseCase(repository);

        allTasksLiveData = getTasksUseCase.execute();
    }

    public LiveData<List<Task>> getAllTasks() {
        return allTasksLiveData;
    }

    public Task getTaskById(int id) {
        return repository.getTaskById(id);
    }

    /**
     * ADD Task + deadline
     * Giữ nguyên logic cũ, chỉ bổ sung notesJson & tablesJson (để dọn chỗ cho 2 feature mới).
     */
    public void addTask(String title, String description, long deadline) {

        Task task = new Task(
                title,
                description,
                System.currentTimeMillis(), // createdAt
                deadline,
                "",   // notesJson — sẽ dùng cho feature Quick Notes
                ""    // tablesJson — sẽ dùng cho feature Tables
        );

        addTaskUseCase.execute(task);
    }

    /**
     * UPDATE Task + deadline
     * ❗ Fix lỗi: phải giữ nguyên notesJson và tablesJson (để không bị mất block file)
     */
    public void updateTask(Task task, String newTitle, String newDesc, long deadline) {

        task.setTitle(newTitle);
        task.setDescription(newDesc);
        task.setDeadline(deadline);

        // notesJson đã được set từ Activity trước khi gọi updateTask()
        // không được reset hoặc bỏ qua

        repository.updateTask(task);
    }

    public void deleteTask(Task task) {
        deleteTaskUseCase.execute(task);
    }

    /**
     * Toggle completed
     */
    public void toggleCompleted(Task task, boolean done) {
        task.setCompleted(done);
        repository.updateCompleted(task, done);
    }

    /**
     * NEW — Lấy task theo ngày
     */
    public LiveData<List<Task>> getTasksByDate(long start, long end) {
        return getTasksByDateUseCase.execute(start, end);
    }
}

File: app/src/main/java/com/example/personaltaskmanager/features/task_manager/viewmodel/TaskViewModel.kt
package com.example.personaltaskmanager.features.task_manager.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import com.example.personaltaskmanager.features.task_manager.data.model.Task
import com.example.personaltaskmanager.features.task_manager.data.repository.TaskRepository

/**
 * ViewModel cho Task.
 * Giữ nguyên code cũ, chỉ bổ sung các hàm cần thiết cho UI.
 */
class TaskViewModel(app: Application) : AndroidViewModel(app) {

    private val repo = TaskRepository(app)

    val allTasks: LiveData<List<Task>> = repo.getAllTasks()

    fun addTask(title: String, desc: String, deadline: Long) {
        val t = Task(
            title,
            desc,
            System.currentTimeMillis(),
            deadline,
            "",
            "",
            0   // userId được Repository tự gán
        )
        repo.addTask(t)
    }

    fun updateTask(task: Task) {
        repo.updateTask(task)
    }

    /**
     * Update có truyền title/desc/deadline (TaskDetailActivity + Workspace cần)
     */
    fun updateTask(task: Task, title: String, desc: String, deadline: Long) {
        task.title = title
        task.description = desc
        task.deadline = deadline
        repo.updateTask(task)
    }

    fun deleteTask(task: Task) {
        repo.deleteTask(task)
    }

    fun getTasksByDate(start: Long, end: Long): LiveData<List<Task>> {
        return repo.getTasksByDate(start, end)
    }

    /**
     * Bổ sung để TaskDetailActivity, WorkspaceActivity, TaskInfoFragment chạy đúng.
     */
    fun getTaskById(taskId: Int): Task? {
        return repo.getTaskById(taskId)
    }

    fun toggleCompleted(task: Task, done: Boolean) {
        task.isCompleted = done
        repo.updateTask(task)
    }
}
